kie-server-parent/kie-server-services/kie-server-services-task-assigning/kie-server-services-task-assigning-planning/src/main/java/org/kie/server/services/taskassigning/planning/data/LabelValueExtractorRegistry.java

```diff
@@ -0,0 +1,82 @@
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Optional;
+import java.util.ServiceLoader;
+import java.util.Set;
+import java.util.function.BiConsumer;
+
+import org.kie.server.api.model.taskassigning.data.LabelValueExtractor;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+public class LabelValueExtractorRegistry {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(LabelValueExtractorRegistry.class);
+
+    private static final LabelValueExtractorRegistry INSTANCE = new LabelValueExtractorRegistry();
+
+    private Map<Class<?>, Set<LabelValueExtractor<?>>> extractorsBySourceType = new HashMap<>();
+
+    private LabelValueExtractorRegistry() {
+        registerExtractors(getClass().getClassLoader());
+    }
+
+    public static LabelValueExtractorRegistry getInstance() {
+        return INSTANCE;
+    }
+
+    public synchronized void registerExtractors(ClassLoader classLoader) {
+        final ServiceLoader<LabelValueExtractor> availableExtractors = ServiceLoader.load(LabelValueExtractor.class, classLoader);
+        for (LabelValueExtractor extractor : availableExtractors) {
+            final Set<LabelValueExtractor<?>> typeExtractors = extractorsBySourceType.computeIfAbsent(extractor.getType(), key -> new HashSet<>());
+            final LabelValueExtractor previousExtractor = typeExtractors.stream()
+                    .filter(previous -> previous.getLabelName().equals(extractor.getLabelName()))
+                    .findFirst().orElse(null);
+            if (previousExtractor != null) {
+                if (previousExtractor.getPriority() < extractor.getPriority()) {
+                    LOGGER.info("LabelValueExtractor with [previousType: {}, previousLabelName:{}, previousPriority: {}], will be replaced with -> [type:{}, labelName:{}, priority: {}]",
+                                previousExtractor.getType(), previousExtractor.getLabelName(), previousExtractor.getPriority(),
+                                extractor.getType(), extractor.getLabelName(), extractor.getPriority());
+                    typeExtractors.remove(previousExtractor);
+                    typeExtractors.add(extractor);
+                }
+            } else {
+                typeExtractors.add(extractor);
+            }
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    public <T> Optional<Set<LabelValueExtractor<T>>> getLabelValueExtractors(Class<T> sourceType) {
+        return (Optional<Set<LabelValueExtractor<T>>>) Optional.ofNullable((T) extractorsBySourceType.get(sourceType));
+    }
+
+    public <T> void applyLabelValueExtractors(Class<T> sourceType, T source, BiConsumer<String, Set<Object>> extractedValuesConsumer) {
+        final Optional<Set<LabelValueExtractor<T>>> extractorsSet = getLabelValueExtractors(sourceType);
+        extractorsSet.ifPresent(extractors -> extractors.forEach(valueExtractor -> {
```

Comment:
I wonder if this code can throw `ConcurrentModificationException`:
In the loop, another thread calls `registerExtractors`. As a result, some extractors may be removed.

Safe approach could be e.g. making `getLabelValueExtractors` return a defensive copy of the original `Set`.

Reply:
Let's say that this could happen in very a "theoretically" worst case, but in practice will never happen. Since the extractors are "configurable" but will never change at run-time or processing time. Instead they are laded at the extension startup as part of all the "planning related" stuff and will never change again.
I didn't add any additional sync mechanism for performance purposes given the statement above.

Reply:
If they are not supposed to be changed after startup, we can make the `registerExtractors` method private to enforce that and even remove the `synchronized` keyword from there.

Reply:
what I'm saying is that registrations can be made as part of the extension initialization. During this time the LabelValueExtractors will never be consumed. Finally only after the extension was initialized, and all potential registration was already produced, the extractors might start being used.
I think current format is ok.